frontend/src/metabase/public/containers/PublicQuestion/PublicQuestion.tsx

```diff
@@ -0,0 +1,245 @@
+import { EmbedFrame } from "metabase/public/components/EmbedFrame";
+import { useEmbedFrameOptions } from "metabase/public/hooks";
+import QueryDownloadWidget from "metabase/query_builder/components/QueryDownloadWidget";
+import { setErrorPage } from "metabase/redux/app";
+import { addParamValues, addFields } from "metabase/redux/metadata";
+import { getMetadata } from "metabase/selectors/metadata";
+import {
+  PublicApi,
+  EmbedApi,
+  setPublicQuestionEndpoints,
+  setEmbedQuestionEndpoints,
+  maybeUsePivotEndpoint,
+} from "metabase/services";
+import { PublicMode } from "metabase/visualizations/click-actions/modes/PublicMode";
+import Visualization from "metabase/visualizations/components/Visualization";
+import Question from "metabase-lib/v1/Question";
+import { getCardUiParameters } from "metabase-lib/v1/parameters/utils/cards";
+import { getParameterValuesBySlug } from "metabase-lib/v1/parameters/utils/parameter-values";
+import { getParametersFromCard } from "metabase-lib/v1/parameters/utils/template-tags";
+import { applyParameters } from "metabase-lib/v1/queries/utils/card";
+import type {
+  Card,
+  VisualizationSettings,
+  Dataset,
+  ParameterId,
+  ParameterValuesMap,
+} from "metabase-types/api";
+
+export const PublicQuestion = ({
+  params: { uuid, token },
+  location,
+}: {
+  location: Location;
+  params: { uuid: string; token: string };
+}) => {
+  const dispatch = useDispatch();
+
+  const metadata = useSelector(getMetadata);
+
+  const [card, setCard] = useState<Card | null>(null);
+  const [result, setResult] = useState<Dataset | null>(null);
+  const [initialized, setInitialized] = useState(false);
+  const [parameterValues, setParameterValues] = useState<ParameterValuesMap>(
+    {},
+  );
+
+  useMount(async () => {
+    if (uuid) {
+      setPublicQuestionEndpoints(uuid);
+    } else if (token) {
+      setEmbedQuestionEndpoints(token);
+    }
+
+    try {
+      let card;
+      if (token) {
+        card = await EmbedApi.card({ token });
+      } else if (uuid) {
+        card = await PublicApi.card({ uuid });
+      } else {
+        throw { status: 404 };
+      }
+
+      if (card.param_values) {
+        await dispatch(addParamValues(card.param_values));
+      }
+      if (card.param_fields) {
+        await dispatch(addFields(card.param_fields));
+      }
+
+      const parameters = getCardUiParameters(
+        card,
+        metadata,
+        {},
+        card.parameters || undefined,
+      );
+      const parameterValuesById = getParameterValuesByIdFromQueryParams(
+        parameters,
+        location.query,
+      );
+
+      setCard(card);
+      setParameterValues(parameterValuesById);
+    } catch (error) {
+      console.error("error", error);
+      dispatch(setErrorPage(error));
+    }
+  });
+
+  const setParameterValue = async (parameterId: ParameterId, value: any) => {
+    setParameterValues({
+      ...parameterValues,
```

Comment:
It's recommended to use the updater function when updating values [based on the previous state.](https://react.dev/reference/react/useState#updating-state-based-on-the-previous-state)

It's not obvious here, but it's necessary when you're updating this multiple times (as mentioned in the React doc) or doing this inside a memoized callback, otherwise, `parameterValues` would become a dependency too, and would trigger more updates than necessary.

```ts
setParameterValues(parameterValuse => ....)
```